unity/Assets/Maroon/reusablePrefabs/NewRuler/Scripts/UIItemDragHandlerUnrestricted.cs

```diff
@@ -25,33 +35,58 @@ public void OnBeginDrag(PointerEventData eventData)
         public void OnDrag(PointerEventData eventData)
         {
             var screenPoint = Input.mousePosition;
-            var finish = parentCanvas.worldCamera.ScreenToWorldPoint(screenPoint);
+            var finish = screenPoint;
             
-            if (!is2D)
+            var ray = Camera.main.ScreenPointToRay(Input.mousePosition);
+
+            if (is2D)
                 finish.z = 0f;
 
+            _objectInsideBoundaries = CheckIfObjectIsInBounds();
             _item.transform.position = finish;
         }
 
         public void OnEndDrag(PointerEventData eventData)
         {
             var ray = Camera.main.ScreenPointToRay(Input.mousePosition);
-            var hits = Physics.RaycastAll(ray, 1000f);
+            var hits = Physics.RaycastAll(ray, 100f);
 
-            foreach (var hit in hits)
+            if (_objectInsideBoundaries)
             {
-                ShowObject(hit.point, hit.transform.parent);
+                if(hits.Length != 0) 
+                    ShowObject(hits.First().point, hits.First().transform.parent);
             }
-
+            
             Destroy(_item);
         }
-
+        
         protected virtual void ShowObject(Vector3 position, Transform parent)
         {
             if (generatedObject == null) return;
 
             generatedObject.transform.position = position;
             generatedObject.SetActive(true);
         }
+
+        private bool CheckIfObjectIsInBounds()
+        {
+
+            if (maxBoundary != null && minBoundary != null)
+            {
+                var maxBoundPosition = maxBoundary.transform.position;
+                var minBoundPosition = minBoundary.transform.position;
+                var ray = Camera.main.ScreenPointToRay(_item.transform.position);
+                var pt = ray.GetPoint(20);
+
+                if (pt.x >= Mathf.Max(minBoundPosition.x, maxBoundPosition.x) || pt.x <= Mathf.Min(minBoundPosition.x, maxBoundPosition.x))
+                    return false;
+                if (pt.y >= Mathf.Max(minBoundPosition.y, maxBoundPosition.y) || pt.y <= Mathf.Min(minBoundPosition.y, maxBoundPosition.y))
+                    return false;
+                if (!is2D && _item.transform.position.z >= Mathf.Max(minBoundPosition.z, maxBoundPosition.z))
```

Comment:
correct me if I'm wrong, but shouldn't this also check transform.position.z against Mathf.Min